### PR TITLE
Fix saving hyperparameters in a composition where parent is not a LM or LDM

### LINKED ISSUE #13181: Unexpected behavior of .save_hyperparameters() for same parameter name
## 🐛 Bug

When a pytorch_lightning model is created as a parameter for a class, and they contain parameters with same name, the original class's value overrides the model's value when calling `.save_hyperparameters()`. Besides, it also saves all parameters for the wrapper class.

### To Reproduce

I created a BoringParent class (not actually a parent) that contains a param named `same_arg` with a default value `parent`, and inside the init function I created a Boring_saving_hprams object that also contains a param named `same_arg` and I call `save_hyperparameters` to save the params for this pytorch_lightning module. However, the value of `same_arg` is overriden by the default BoringParent class's value.
Please see the colab notebook for details.
https://colab.research.google.com/drive/1De82M7GEzgtc18UTSRHJOV_nNnIJrji6?usp=sharing

### Expected behavior

I'm expecting the model should save its own value even when there're param name conflicts. And in this case the class BoringParent is not actually the parent, it's just the object holding this model, we shouldn't iterate over its parameters. I also just want it to save the parameters for this pytorch_lightning modle only, not parameters from the class that holds this model.

### Environment

* CUDA:
	- GPU:
		- Tesla T4
	- available:         True
	- version:           11.3
* Packages:
	- numpy:             1.21.6
	- pyTorch_debug:     False
	- pyTorch_version:   1.11.0+cu113
	- pytorch-lightning: 1.6.3
	- tqdm:              4.64.0
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.7.13
	- version:           #1 SMP Sun Apr 24 10:03:06 PDT 2022

### Additional context

Inside the save_hyperparameters function, it's updating the init_args here so the later args just override the previous one:
image
And inside `collect_init_args` it's recursively checking the params passed in, which is not the correct order for my case. I think that we could check whether there's a real inheritance relationship between classes and if not we don't add the param inside the path_args.

Also let me know if you want me to draft a PR. Thanks!

cc @carmocca @justusschock @awaelchli @borda @ananthsub @ninginthecloud @jjenniferdai @rohitgr7

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #13181
Closes: #13202

Look at this code:

```py
    class Child(LightningModule):
        def __init__(self, same_arg):
            super().__init__()
            self.save_hyperparameters()

    class Parent: 
        def __init__(self, same_arg="parent_default", other_arg="other"):
            self.child = Child(same_arg="cocofruit")

    parent = Parent()
    print(parent.child.hparams)
```

Note: There is a LightningModule composed inside of a non-LightningModule class. 

**Q: At the call `self.save_hyperparameters()`, which parameters do we excpect to get saved?**
A: Only the ones from the child LightningModule. But on master, it collects also the ones from the parent class. 

**Q: But Adrian, collecting more parameters is always better. The more the better!**
A: No. 1) You could coincidentally have two arguments with the same name in two different classes. So there is a collision and in master, one overwrites the other. Bad! 2) We save hyperparameters not just for the sake of saving them. We want to save exactly the ones needed to reconstruct the object when loading `ChildInComposition.load_from_checkpoint`. No more, no less.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @borda @justusschock @awaelchli @ninginthecloud @rohitgr7 @otaj @akihironitta

### CODE REVIEW

File: src/pytorch_lightning/utilities/parsing.py [hunk 1]
```diff
@@ -181,13 +185,13 @@ def collect_init_args(
     if not isinstance(frame.f_back, types.FrameType):
         return path_args
 
-    if "__class__" in local_vars:
+    if "__class__" in local_vars and issubclass(local_vars["__class__"], classes):
```
Comment (awaelchli) [inline src/pytorch_lightning/utilities/parsing.py:188]: Reviewers: This function `collect_init_args` is recursive, it will go up the stack frames until it hits a class's `__init__`. But this init could be inside of another init frame. On master, we would just continue iterating. Here, I'm making sure that we are only doing shenanigans when the init belongs to a subclass of LM or LDM.

File: src/pytorch_lightning/utilities/parsing.py [hunk 2]
```diff
@@ -162,14 +162,18 @@ def get_init_args(frame: types.FrameType) -> Dict[str, Any]:
 
 
 def collect_init_args(
-    frame: types.FrameType, path_args: List[Dict[str, Any]], inside: bool = False
+    frame: types.FrameType,
+    path_args: List[Dict[str, Any]],
+    inside: bool = False,
+    classes: Tuple[Type, ...] = (object,),
```
Comment (carmocca) [inline src/pytorch_lightning/utilities/parsing.py:168]: Initializing the tuple to contain a class looks like a code smell to me.

You could also do:

```py
classes = (,)
...

if "__class__" in local_vars and not classes or issubclass(local_vars["__class__"], classes):
    ...
```

**Review by carmocca — APPROVED** [review]

File: src/pytorch_lightning/utilities/parsing.py (hunk 2 above)
Comment (awaelchli) [reply src/pytorch_lightning/utilities/parsing.py]: What about it is a code smell? 
An empty tuple expresses that there are no classes we should search for, but on the contrary, we interpret it as looking for all of them. The special casing only makes it more difficult to understand the code now.

Comment (carmocca) [reply src/pytorch_lightning/utilities/parsing.py]: I guess it's not really a problem because (1) tuples are immutable and (2) it contains a class type which is a singleton anyways. But in general this pattern is discouraged because all function calls will share the same instance

```python
def foo(arg=tuple()):
    return arg

a = foo()
b = foo()
print(a is b)
```

So it's up to you. I prefer the explicit check but feel free to revert. It's got my approval both ways :smile:

**Review by otaj — APPROVED** [review]

File: src/pytorch_lightning/CHANGELOG.md [hunk 3]
```diff
@@ -70,6 +70,10 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed epoch-end logging results not being reset after the end of the epoch ([#14061](https://github.com/Lightning-AI/lightning/pull/14061))
 
 
+- Fix saving hyperparameters in a composition where the parent class is not a `LightningModule` or `LightningDataModule` ([#14151](https://github.com/Lightning-AI/lightning/pull/14151))
```
Comment (rohitgr7) [inline src/pytorch_lightning/CHANGELOG.md:73]:
```suggestion
- Fixed saving hyperparameters in a composition where the parent class is not a `LightningModule` or `LightningDataModule` ([#14151](https://github.com/Lightning-AI/lightning/pull/14151))
```

**Review by rohitgr7 — APPROVED** [review]
